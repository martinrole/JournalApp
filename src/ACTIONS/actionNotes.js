import { addDoc, collection, deleteDoc, doc, updateDoc } from "@firebase/firestore"
import Swal from "sweetalert2"
import { baseDatos } from "../FIREBASE/firebaseConfig"
import { fileUploadHelper } from "../HELPERS/fileUpload"
import { loadNotes } from "../HELPERS/loadNotes"
import { types } from "../TYPES/types"


export const newNoteAction = () => {
    return async ( dispatch, getState ) => {
        const { uid } = getState().auth

        const newNote = {
            titulo: '',
            texto: '',
            fecha: new Date().getTime()
        }

        const docRef = await addDoc( collection(baseDatos, `${uid}/journal/notes`), newNote )
        //console.log('docRef: ', docRef);
        
        dispatch( activeNote(docRef.id, newNote) )
        dispatch( addNewNote(docRef.id, newNote) )
    }
}

export const loadNotesAction = ( uid ) => {
    return async ( dispatch ) => {
        const notes = await loadNotes( uid )
        dispatch( setNotes( notes ) )
    }
}

export const saveNoteAction = ( note ) => async ( dispatch, getState ) => {
        
    const { uid } = getState().auth

    //En los 3 puntos queda el resto de propiedades de la nota, y así se puede extraer el id para no repetirlo
    const {id, imgUrl = null, ...notaFinal } = note

    //Asi se envia a Firebase en la versiòn beta 9
    const noteRef = doc(baseDatos, `${uid}/journal/notes/${note.id}`)
    await updateDoc( noteRef, { imgUrl, ...notaFinal })

    dispatch( refreshNote( note.id, note ))
    Swal.fire('Saved', note.titulo, 'success' )
}

export const deleteNoteAction = ( id ) => async ( dispatch, getState ) => {

    const { uid } = getState().auth

    const  noteRef = doc(baseDatos, `${uid}/journal/notes/${id}`)
    await deleteDoc(noteRef)

    dispatch( deleteNote( id ) )
}

export const fileUpload = ( archivo ) => async ( dispatch, getState ) => {
    
    const activeNote = getState().notes.activeNote

    Swal.fire({
        title: 'Uploading...',
        text: 'Please wait...',
        allowOutsideClick: false,
        showConfirmButton: false,
        willOpen: () => {
            Swal.showLoading()
        }
    })

    
    const fileUrl = await fileUploadHelper( archivo )
    activeNote.imgUrl = fileUrl
    
    dispatch( saveNoteAction( activeNote ) )
    Swal.close()

}

export const addNewNote = ( id, nota ) => ({
    type: types.notesAdd,
    payload: { id, ...nota }
})

export const refreshNote = ( id, nota ) => ({
    type: types.notesUpdate,
    payload: { id, nota: { id, ...nota } }
})


export const activeNote = (id, nota) => ({
    type: types.notesActive,
    payload: { id, ...nota }
})

export const setNotes = ( nota ) => ({
    type: types.notesLoad,
    payload: nota
})

export const deleteNote = ( id ) => ({
    type: types.notesDelete,
    payload: id
})

export const clearNotes = () => ({
    type: types.notesClear
})
