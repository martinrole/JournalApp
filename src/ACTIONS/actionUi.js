import { types } from "../TYPES/types"

export const actionSetError = ( error ) => ({
    type: types.uiSetError,
    payload: error
})

export const actionRemoveError = () => ({
    type: types.uiRemoveError
})


export const actionLoading = (type) => ({
    type: type
})

