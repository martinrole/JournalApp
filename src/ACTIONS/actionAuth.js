import { autorizacion, googleAuth } from "../FIREBASE/firebaseConfig";
import { signInWithPopup, createUserWithEmailAndPassword, updateProfile, signInWithEmailAndPassword, signOut } from 'firebase/auth';
import { types } from "../TYPES/types"
import { actionLoading, actionSetError } from "./actionUi";
import Swal from 'sweetalert2'
import { clearNotes } from "./actionNotes";

//Middleware:
export const actionLoginAsync = ( correo, contrasena ) => {
    //Esta funciÃ³n va a retornar un Callback:
    //el dispatch lo va  a entregar el thunk instalado
    return (dispatch) => {

        dispatch( actionLoading(types.uiLoadingOn) )

        return signInWithEmailAndPassword( autorizacion, correo, contrasena )
            .then( ({user}) => {
                dispatch( actionLogin( user.uid, user.displayName ) )
                dispatch( actionLoading(types.uiLoadingOff) )
            })
            .catch ( error => {
                dispatch( actionSetError(error.code) )
                Swal.fire('Error', error.code, 'error')
                dispatch( actionLoading(types.uiLoadingOff) )
            }) 
            
        }
}



export const actionRegisterAsync = ( nombre, correo, contrasena ) => {
    return ( dispatch ) => {

        return createUserWithEmailAndPassword( autorizacion, correo, contrasena )
            .then( async ( { user } ) => {
                await updateProfile( user, { displayName: nombre } )
                
                dispatch(
                    actionLogin( user.uid, user.displayName)
                )
            })
            .catch ( error => {
                dispatch( actionSetError(error.code) )
                Swal.fire('Error', error.code, 'error')
            })  
    }
}

export const actionGoogleLogin = () => {
    return ( dispatch ) => {
        signInWithPopup( autorizacion, googleAuth )
            .then( ( { user } ) => {
                dispatch(
                    actionLogin( user.uid, user.displayName)
                )

            })     
    }
}

export const actionLogOutAsync = () => {
    return async ( dispatch ) => {
        await signOut(autorizacion)

        dispatch( actionLogOut() )
        dispatch( clearNotes() )
    }
}

//Una forma de hacerlo:
// export const actionAuth = (uid, displayName) => {
//     return {
//         type: types.login,
//         payload: {
//             uid,
//             displayName
//         }
//     }
// }


//Segunda forma de hacerlo:
export const actionLogin = (uid, displayName) => ({  
    type: types.login,
    payload: {
        uid,
        displayName
    }
})

export const actionLogOut = () => ({
    type: types.logout
})