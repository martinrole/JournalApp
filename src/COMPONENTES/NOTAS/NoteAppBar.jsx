import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { fileUpload, saveNoteAction } from '../../ACTIONS/actionNotes'

export const NoteAppBar = () => {

    const dispatch = useDispatch()
    const { activeNote } = useSelector(state => state.notes)

    const saveClick = () => {
        dispatch( saveNoteAction( activeNote ) )
    }

    const pictureClick = () => {
        // document.getElementById('inputFile').click()
        document.querySelector('#inputFile').click()
    }

    const handleFileChange = (e) => {
        const file = e.target.files[0]

        if (file) {
            dispatch( fileUpload(file) )
        }

    }

    return (
        <div className="notes__appbar">
            <span>28 de Agosto de 2020</span>

            <input 
                id="inputFile"
                name="fileNote"
                type="file" 
                style={ { display: 'none' } } 
                onChange={ handleFileChange }
            />

            <div>
                <button className="btn" onClick={ pictureClick }>Picture</button>
                <button className="btn" onClick={ saveClick }>Save</button>

            </div>
        </div>
    )
}
