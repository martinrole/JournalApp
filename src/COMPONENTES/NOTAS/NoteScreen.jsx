import React, { useEffect, useRef } from 'react'
import { useSelector } from 'react-redux'
import { useForm } from '../../HOOKS/useForm'
import { NoteAppBar } from './NoteAppBar'
import { useDispatch} from 'react-redux'
import { activeNote, deleteNoteAction } from '../../ACTIONS/actionNotes'

export const NoteScreen = () => {

    const { activeNote: nota } = useSelector(state => state.notes) 
    const[ formValues, handleInputChange, reset ] = useForm( nota )
    const { titulo, texto } = formValues

    const activeId = useRef( nota.id )
    const dispatch = useDispatch()
    
    useEffect(() => {
       
        if ( nota.id !== activeId.current ) {
            //ExplicaciÃ³n de esto video 266 Minuto 4
            reset( nota )
            activeId.current = nota.id
        }

    }, [reset, nota])


    useEffect(() => {

        dispatch( activeNote( formValues.id, { ...formValues } ) )
    }, [formValues, dispatch])

    const handleDelete = () => {
        dispatch( deleteNoteAction( formValues.id ) )
    }

    return (
        <div className="notes__main-content">
            <NoteAppBar />

            <div className="notes__content">
                <input 
                    type="text"
                    name="titulo" 
                    placeholder="Write title" 
                    className="notes__title-input" 
                    autoComplete="off"
                    value={ titulo }
                    onChange={ handleInputChange }
                />
                <textarea 
                    placeholder="What happened today?"
                    name="texto" 
                    className="notes__textarea" 
                    cols="30" 
                    rows="10"
                    value={ texto }
                    onChange={ handleInputChange } 
                ></textarea>

                {
                    ( nota.imgUrl ) 
                    && (
                        <div className="notes__image">
                            <img src={ nota.imgUrl } alt="img note" />
                        </div>
                    )
                }

            </div>

            <button className="btn btn-danger" onClick={ handleDelete }>Delete Note</button>
        </div>
    )
}
