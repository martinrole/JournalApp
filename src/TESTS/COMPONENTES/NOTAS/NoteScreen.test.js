import configureStore from 'redux-mock-store' //ES6 modules
import thunk from 'redux-thunk';
import '@testing-library/jest-dom'
import { mount } from 'enzyme';
import { Provider } from "react-redux";
import { NoteScreen } from '../../../COMPONENTES/NOTAS/NoteScreen';
import { activeNote } from '../../../ACTIONS/actionNotes';

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initState = {
    auth: {},
    ui: {
        cargando: false
    },
    notes: {
        notes: [],
        activeNote: {
            id: 1234,
            titulo: 'Prueba: titulo de la nota',
            texto: 'Texto de la prueba de la nota',
            fecha: 0
        }
    }
}

let store = mockStore( initState )
store.dispatch = jest.fn()


//Simula la ejecución de la función
jest.mock('../../../ACTIONS/actionNotes', () => ({
    activeNote: jest.fn(),
}))

const wrapper = mount( 
    <Provider store={ store }>
        <NoteScreen />
    </Provider>
)

describe('Pruebas en componente NoteScreen', () => {

    test('Debe mostrar el componente correctamente', () => {
        expect( wrapper ).toMatchSnapshot()
    })

    test('Debe de activar el estado activeNote', () => {
        
        wrapper.find('textarea[name="texto"]').simulate('change', {
            target: {
                name: 'texto',
                value: 'Debe cambiar el texto por este'
            }
        })

        expect( activeNote ).toHaveBeenCalledTimes(1)
        expect( activeNote ).toHaveBeenCalledWith(
            1234,
            {
                id: 1234,
                titulo: 'Prueba: titulo de la nota',
                texto: 'Debe cambiar el texto por este',
                fecha: 0
            }
        )
    })
})