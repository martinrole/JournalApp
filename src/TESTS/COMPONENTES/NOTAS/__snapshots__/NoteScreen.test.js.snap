// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente NoteScreen Debe mostrar el componente correctamente 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      28 de Agosto de 2020
    </span>
    <input
      id="inputFile"
      name="fileNote"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="titulo"
      onChange={[Function]}
      placeholder="Write title"
      type="text"
      value="Prueba: titulo de la nota"
    />
    <textarea
      className="notes__textarea"
      cols="30"
      name="texto"
      onChange={[Function]}
      placeholder="What happened today?"
      rows="10"
      value="Texto de la prueba de la nota"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete Note
  </button>
</div>
`;
